File: src/flood_fill_pkg/flood_fill_pkg/flood_fill.py
from collections import deque

'''
0 = free path
1 = obstacle
2 = goal
'''

def flood_fill(start_point, grid):
    start = (start_point.x, start_point.y, start_point.z)
    
    queue = deque([start])
    visited = set([start])  # Contains nodes that have been processed
    deltas = [-1, 0, 1]

    while queue:
        x, y, z = queue.popleft()
        current_value = grid[x][y][z]

        for dx in deltas:
            for dy in deltas:
                for dz in deltas:
                    if dx == 0 and dy == 0 and dz == 0:
                        continue
                    
                    nx, ny, nz = x + dx, y + dy, z + dz
                    neighbor = (nx, ny, nz)
                    
                    if neighbor in visited:
                        continue

                    if is_point_valid(neighbor, grid):
                        if grid[nx][ny][nz] == 2:
                            print("Path found to a goal at:", neighbor)
                        elif grid[nx][ny][nz] == 0:
                            queue.append(neighbor)
                            grid[nx][ny][nz] = current_value + 1

                    visited.add(neighbor)
    return grid

def reverse_flood_fill2(start_point, goal, grid):
    if (grid[start_point[0]][start_point[1]][start_point[2]]) == 1:
        print("The start is barrier!")
        return
    # goal = (goal_point.x, goal_point.y, goal_point.z)
    grid[goal[0]][goal[1]][goal[2]] = 2
    queue = deque([goal])
    visited = set([goal])  # Contains nodes that have been processed
    deltas = [-1, 0, 1]

    while queue:
        x, y, z = queue.popleft()
        current_value = grid[x][y][z]

        for dx in deltas:
            for dy in deltas:
                for dz in deltas:
                    if dx == 0 and dy == 0 and dz == 0:
                        continue
                    
                    nx, ny, nz = x + dx, y + dy, z + dz
                    neighbor = (nx, ny, nz)
                    
                    if neighbor in visited:
                        continue

                    if is_point_valid(neighbor, grid) and grid[nx][ny][nz] == 0:
                        grid[nx][ny][nz] = current_value + 1
                        queue.append(neighbor)

                    visited.add(neighbor)

    if grid[start_point[0]][start_point[1]][start_point[2]] > 2:
        print(f"Path found to the start from the goal with a length of { grid[start_point[0]][start_point[1]][start_point[2]] - 2}")
    else:
        print(f"Path not found! start_value: { grid[start_point[0]][start_point[1]][start_point[2]]}" )
    return grid

def is_point_valid(point, grid) -> bool:
    x, y, z = point
    if 0 <= x < len(grid):
        if 0 <= y < len(grid[x]):
            if 0 <= z < len(grid[x][y]):
                cell_value = grid[x][y][z]
                return cell_value == 0  # Only visit if cell value is 0 (unvisited)
    return False

def is_point_valid_path(point, grid) -> bool:
    x, y, z = point
    if 0 <= x < len(grid):
        if 0 <= y < len(grid[x]):
            if 0 <= z < len(grid[x][y]):
                cell_value = grid[x][y][z]
                return cell_value != 1  # Only visit if cell value is not 1 (barrier)
    return False

def find_path(starting_point, goal_point, grid):
    path = []
    current_position = starting_point  # this should be a tuple (x, y, z)
    deltas = [-1, 0, 1]

    while current_position != goal_point:
        x, y, z = current_position
        path.append(current_position)

        min_neighbor = None
        min_neighbor_value = float('inf')
        
        for dx in deltas:
            for dy in deltas:
                for dz in deltas:
                    if dx == 0 and dy == 0 and dz == 0:
                        continue

                    nx, ny, nz = x + dx, y + dy, z + dz
                    neighbor = (nx, ny, nz)
                    # we choose the lowest value from neighbors
                    if is_point_valid_path(neighbor, grid) and grid[nx][ny][nz] < min_neighbor_value:
                        min_neighbor_value = grid[nx][ny][nz]
                        min_neighbor = neighbor

        # If no valid neighbor was found, break out of the loop.
        # This might happen if there's no valid path to the goal
        if min_neighbor is None:
            break

        current_position = min_neighbor

    return path

def reverse_flood_fill(start_point, goal, grid):
    if grid[start_point[0]][start_point[1]][start_point[2]] == 1:
        print("The start is a barrier!")
        return grid
    
    x_len, y_len, z_len = len(grid), len(grid[0]), len(grid[0][0])
    grid[goal[0]][goal[1]][goal[2]] = 2
    queue = deque([goal])
    
    # Use deltas without the (0, 0, 0) combination.
    deltas = [-1, 0, 1]
    combinations = [(dx, dy, dz) for dx in deltas for dy in deltas for dz in deltas if dx or dy or dz]

    while queue:
        x, y, z = queue.popleft()
        current_value = grid[x][y][z]

        for dx, dy, dz in combinations:
            nx, ny, nz = x + dx, y + dy, z + dz

            # Use inlined boundary checks.
            if 0 <= nx < x_len and 0 <= ny < y_len and 0 <= nz < z_len and grid[nx][ny][nz] == 0:
                grid[nx][ny][nz] = current_value + 1
                queue.append((nx, ny, nz))
    
    start_value = grid[start_point[0]][start_point[1]][start_point[2]]
    if start_value > 2:
        print(f"Path found to the start from the goal with a length of {start_value - 2}")
    else:
        print(f"Path not found! start_value: {start_value}")
    
    return grid


def reverse_flood_fill_4(start_point, goal, grid):
    if grid[start_point[0]][start_point[1]] == 1:
        print("The start is a barrier!")
        return grid
    
    x_len, y_len, z_len = len(grid), len(grid[0]), len(grid[0][0])
    grid[goal[0]][goal[1]][goal[2]] = 2
    queue = deque([goal])
    
    # Define the 4 cardinal direction offsets.
    directions = [(1, 0, 0), (-1, 0, 0), (0, 1, 0), (0, -1, 0), (0, 0, 1), (0, 0, -1)]

    while queue:
        x, y, z = queue.popleft()
        current_value = grid[x][y][z]

        for dx, dy, dz in directions:
            nx, ny, nz = x + dx, y + dy, z + dz

          # Use inlined boundary checks.
            if 0 <= nx < x_len and 0 <= ny < y_len and 0 <= nz < z_len and grid[nx][ny][nz] == 0:
                grid[nx][ny][nz] = current_value + 1
                queue.append((nx, ny, nz))
    
    start_value = grid[start_point[0]][start_point[1]][start_point[2]]
    if start_value > 2:
        print(f"Path found to the start from the goal with a length of {start_value - 2}")
    else:
        print(f"Path not found! start_value: {start_value}")
    
    return grid

File: src/flood_fill/FloodFill.cpp
// #include <rclcpp/rclcpp.hpp>
#include <iostream>
#include <vector>
#include <queue>

struct Point {
    int x, y, z;
    // Point(int _x, int _y, int _z) : x(_x), y(_y), z(_z) {}
};

void reverse_flood_fill(Point start, Point goal, std::vector<std::vector<std::vector<int>>> map)
{
    // Define the 6 face neighbor offsets.
    int directions[6][3] = {
        {1, 0, 0}, {-1, 0, 0},
        {0, 1, 0}, {0, -1, 0},
        {0, 0, 1}, {0, 0, -1}
    };

    int x_len = map.size();
    int y_len = map[0].size();
    int z_len = map[0][0].size();
    
    int deltas[3] = {-1, 0, 1};

    std::queue<Point> q;
    q.push(goal);

    while (!q.empty())
    {   
        Point current_point = q.front();
        q.pop();

        int current_value = map[current_point.x][current_point.y][current_point.z];
        // 26 susednost
        for (int dx : deltas)
        {
            for (int dy : deltas)
            {
                for (int dz : deltas)
                {
                    if (dx == 0 && dy == 0 && dz == 0)
                    {
                        continue;
                    } 
                    Point neighbor{current_point.x + dx, current_point.y + dy, current_point.z + dz};

                    if (0 <= neighbor.x && neighbor.x < x_len && 0 <= neighbor.y && neighbor.y < y_len && 0 <= neighbor.z && neighbor.z < z_len && map[neighbor.x][neighbor.y][neighbor.z] == 0) {
                        map[neighbor.x][neighbor.y][neighbor.z] = current_value + 1;
                        q.push(neighbor);
                    }
                }
            }
        }
        // 6 susednost
        // for (int i = 0; i < 6; ++i) {
        //     Point neighbor{current_point.x + directions[i][0], current_point.y + directions[i][1], current_point.z + directions[i][2]};

        //     // Use inlined boundary checks.
        //     if (0 <= neighbor.x && neighbor.x < x_len && 0 <= neighbor.y && neighbor.y < y_len && 0 <= neighbor.z && neighbor.z < z_len && map[neighbor.x][neighbor.y][neighbor.z] == 0) {
        //         map[neighbor.x][neighbor.y][neighbor.z] = current_value + 1;
        //         q.push(neighbor);
        //     }
        // }
    }

    int start_value = map[start.x][start.y][start.z];
    if (start_value > 2) {
        std::cout << "Path found to the start from the goal with a length of " << start_value - 2 << std::endl;
    } else {
        std::cout << "Path not found! start_value: " << start_value << std::endl;
    }
}


int main(int argc, char **argv)
{   
    // Create a 3D map of dimensions 10x10x10 filled with zeros.
    std::vector<std::vector<std::vector<int>>> map(10, std::vector<std::vector<int>>(288, std::vector<int>(366, 0)));

    // Define the start and goal points.
    Point start_point = {0, 0, 0};
    Point goal_point = {9, 9, 9};

    map[goal_point.x][goal_point.y][goal_point.z] = 2;

    // Call the reverse_flood_fill function.
    reverse_flood_fill(start_point, goal_point, map);

    return 0;
}

